src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -967,13 +967,11 @@ private string ReplaceCore(string oldValue, string? newValue, CompareInfo? ci, C
         //
         public string Replace(char oldChar, char newChar)
         {
-            if (oldChar == newChar)
-                return this;
-
-            int firstIndex = IndexOf(oldChar);
-
-            if (firstIndex < 0)
+            int firstIndex;
+            if (oldChar == newChar || (firstIndex = IndexOf(oldChar)) < 0)
```

Comment:
isn't it against the guidelines to perform an assignment inside an `if` ?

Reply:
I don't know, but it gives nice machine code here 😉
It's about collapsing the epilogs for the first checks (`oldChar == newChar`, and `firstIndex < 0`).
Maybe I think too complicated now, but the other option would be using `goto` for this.

Reply:
At the very least it would be good to have a comment calling out the assignment why it is being done here.

Otherwise, at a glance it may look like a potential bug or comparison using `==`

Reply:
But, the below is also the more "natural" pattern and more readable:
```csharp
if (oldChar == newChar)
{
    return this;
}

int firstIndex = IndexOf(oldChar);

if (firstIndex < 0)
{
    return this;
}
```

Ideally the JIT would handle such a pattern "correctly" and optimize it down accordingly.

Reply:
@tannergooding so what's my action here?
* add the comment explaining that the epilogs get collapsed in generated machine-code
* write it more naturally and file an issue for the JIT 
* do both, and write it naturally once the JIT issue is fixed

(I'm leaning towards the last option, for perf-reasons -- except JIT issue will be fixed for .NET 7 😉)

Reply:
I think writing it naturally and filing an issue for the JIT is the best choice and don't expect the cost to be significant here.

If the cost is more significant, then adding a comment calling out the assignment and why as well as filing an issue for the JIT is the next best option.

If the issue is actually being fixed for .NET 7, that's all the more reason to do the first approach.

Reply:
Is the codegen issue tracked by https://github.com/dotnet/runtime/issues/8883?